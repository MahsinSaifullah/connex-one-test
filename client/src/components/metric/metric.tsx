import React from 'react';

export const Metric: React.FC = () => {
  return <div>metric</div>;
};
