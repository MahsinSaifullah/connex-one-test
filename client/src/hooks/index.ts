export * from './time.hook';
export * from './metric.hook';
