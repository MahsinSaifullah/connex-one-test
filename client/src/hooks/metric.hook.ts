import * as React from 'react';
import { getMetrics } from '../services';

export const useMetric = () => {
  const [metrics, setMetrics] = React.useState<string>('');
  const [isLoading, setIsLoading] = React.useState(false);
  const [isError, setIsError] = React.useState(false);
  const metricFetchRef = React.useRef<NodeJS.Timer>();

  const getPromMetrics = async () => {
    try {
      setIsLoading(true);
      const _metrics = await getMetrics();
      setMetrics(_metrics);
    } catch (error) {
      setIsError(true);
    }
    setIsLoading(false);
  };

  React.useEffect(() => {
    getPromMetrics();

    if (metricFetchRef.current) {
      clearInterval(metricFetchRef.current);
    }

    metricFetchRef.current = setInterval(async () => {
      setIsError(false);
      await getPromMetrics();
    }, 30000);

    return () => {
      clearInterval(metricFetchRef.current);
    };
  }, []);

  return {
    metrics,
    isError,
    isLoading,
  };
};
