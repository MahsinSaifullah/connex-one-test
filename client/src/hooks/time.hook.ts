import * as React from 'react';

import { IServerTime, ITimeDifference } from '../models';
import { getServerTime } from '../services';
import { calculateTimeDifference, convertSecondToHourMinSec } from '../utils';

export const useTime = () => {
  const [serverTime, setServerTime] = React.useState<IServerTime>({
    epoch: 0,
  });
  const [timeDiff, setTimeDiff] = React.useState<ITimeDifference>({
    hours: 0,
    minutes: 0,
    seconds: 0,
  });
  const [isLoading, setIsLoading] = React.useState(false);
  const [isError, setIsError] = React.useState(false);
  const timeFetchRef = React.useRef<NodeJS.Timer>();
  const timeDiffRef = React.useRef<NodeJS.Timer>();

  const getTimeDiff = () => {
    const timeDiffFromServer = calculateTimeDifference(serverTime.epoch);
    setTimeDiff(convertSecondToHourMinSec(timeDiffFromServer));
  };

  const getTime = async () => {
    try {
      setIsLoading(true);
      const time = await getServerTime();
      setServerTime(time);
    } catch (error) {
      setIsError(true);
    }

    setIsLoading(false);
  };

  React.useEffect(() => {
    getTime();

    if (timeFetchRef.current) {
      clearInterval(timeFetchRef.current);
    }

    timeFetchRef.current = setInterval(async () => {
      await getTime();
    }, 30000);

    return () => {
      clearInterval(timeFetchRef.current);
    };
  }, []);

  React.useEffect(() => {
    if (timeDiffRef.current) {
      clearInterval(timeDiffRef.current);
    }

    timeDiffRef.current = setInterval(async () => {
      getTimeDiff();
    }, 1000);

    return () => {
      clearInterval(timeDiffRef.current);
    };
  }, [serverTime.epoch]);

  return {
    serverTime,
    timeDiff,
    isLoading,
    isError,
  };
};
