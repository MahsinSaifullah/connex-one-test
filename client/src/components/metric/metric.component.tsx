import * as React from 'react';

import { useMetric } from '../../hooks';
import { Card, Loading, Title } from '../shared';

export const Metric: React.FC = () => {
  const { metrics, isLoading } = useMetric();
  return (
    <div className="flex flex-col justify-center items-center space-y-4">
      <Title title="Prometheus Metrics" />
      <Card className="sm:min-w-[600px] max-w-sm lg:max-w-xl h-[400px] xl:h-[500px] px-4 py-2">
        {isLoading ? (
          <Loading loadingText="Loading" />
        ) : (
          <pre className="text-xs sm:text-sm tracking-widest font-mono font-bold">
            {metrics}
          </pre>
        )}
      </Card>
    </div>
  );
};
