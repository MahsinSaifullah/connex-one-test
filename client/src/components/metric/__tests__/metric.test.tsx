import { render, screen } from '@testing-library/react';

import * as hooks from '../../../hooks';
import { Metric } from '../metric.component';

jest.mock('../../../hooks', () => {
  return {
    __esModule: true,
    ...jest.requireActual('../../../hooks'),
  };
});

const spiedMetricHook = jest.spyOn(hooks, 'useMetric');

describe('Metric Component', () => {
  it('should render loading component if isLoading is true', () => {
    spiedMetricHook.mockReturnValue({
      metrics: '',
      isLoading: true,
      isError: false,
    });

    const { asFragment } = render(<Metric />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render Metric component if isLoading is false', () => {
    spiedMetricHook.mockReturnValue({
      metrics: 'test',
      isLoading: false,
      isError: false,
    });

    const { asFragment } = render(<Metric />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render the correct metrics', () => {
    spiedMetricHook.mockReturnValue({
      metrics: 'test',
      isLoading: false,
      isError: false,
    });

    render(<Metric />);
    expect(screen.getByTestId('metrics').innerHTML).toBe('test');
  });
});
