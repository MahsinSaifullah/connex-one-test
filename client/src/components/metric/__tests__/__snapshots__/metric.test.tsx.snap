// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Metric Component should render Metric component if isLoading is false 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col justify-center items-center space-y-4"
  >
    <h2
      class="text-lg sm:text-2xl font-bold tracking-widest undefined"
    >
      Prometheus Metrics
    </h2>
    <article
      class="rounded shadow-lg bg-white overflow-scroll sm:min-w-[600px] max-w-sm lg:max-w-xl h-[400px] xl:h-[500px] px-4 py-2"
    >
      <pre
        class="text-xs sm:text-sm tracking-widest font-mono font-bold"
        data-testid="metrics"
      >
        test
      </pre>
    </article>
  </div>
</DocumentFragment>
`;

exports[`Metric Component should render loading component if isLoading is true 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col justify-center items-center space-y-4"
  >
    <h2
      class="text-lg sm:text-2xl font-bold tracking-widest undefined"
    >
      Prometheus Metrics
    </h2>
    <article
      class="rounded shadow-lg bg-white overflow-scroll sm:min-w-[600px] max-w-sm lg:max-w-xl h-[400px] xl:h-[500px] px-4 py-2"
    >
      <div
        class="text-2xl animate-pulse undefined"
      >
        Loading ...
      </div>
    </article>
  </div>
</DocumentFragment>
`;
